MAINT: stats: Add custom reprs for transformed distributions

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message.
However, please only include an issue number in the description, not the title,
and please ensure that any code names containing underscores are enclosed in backticks.

Depending on your changes, you can skip CI operations and save time and energy: 
https://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->

#### What does this implement/fix?
<!--Please explain your changes.-->
This PR adds custom reprs for distributions in the new infrastructure produced by:

* shifting and scaling, e.g.`2*X + 1`
* producing a truncated distribution with `stats.truncate`
* creating the distribution for an order_statistic with `stats.order_statistic`.
* Creating a mixture model with `Mixture`

I removed the `repr` override in `TransformedDistribution`, since it doesn't work well when transforming a transformed distribution, and also isn't really a great default. I think the expectation should just be that transformed distributions get their own `repr`.

In addition, I've also taken advantage of `dict`s being uniquely orderd in Python 3.7+ , and made it so the `dict` of parameters for a distribrution are presorted, rather than sorted downstream whenever they are used. I did this in such a way that when a new distribution is created with a transform, the transform specific parameters will appear in the parameter `dict` for the new distribution after the original distribution parameters. This impacted the order of the parameters in old `repr`s, which is why it made it's way into this PR. It does not impact the custom `repr`s which I implemented after making that change, but I still think it's a good change to make. 

Examples for how the reprs look

Shifting and scaling produces an arithmetic expression like this

```pycon
>>> import scipy.stats as stats
>>> X = stats.Normal(mu=0, sigma=1)
>>> Y = 2*X + 1
>>> Y
2.0*Normal(mu=0.0, sigma=1.0) + 1.0
```

For `truncate` and `order_statistic`, the `repr` mimics the function calls that produced the transformed distribution

```pycon
>>> Z = stats.truncate(Y, -1, 1)
>>> Z
truncate(2.0*Normal(mu=0.0, sigma=1.0) + 1.0, lb=-1.0, ub=1.0)
>>> W = stats.order_statistic(Z, r=3, n=10)
>>> W
order_statistic(truncate(2.0*Normal(mu=0.0, sigma=1.0) + 1.0, lb=-1.0, ub=1.0), r=3.0, n=10.0)
```

`Mixture` also replicates the calls that produced the transformed distribution. I've used black style wrapping here to keep the lines from getting too long.

```pycon
>>> M = stats.Mixture([X, Y, Z, W], weights=[0.25, 0.25, 0.25, 0.25])
>>> M
Mixture(
    [
        Normal(mu=0.0, sigma=1.0),
        2.0*Normal(mu=0.0, sigma=1.0) + 1.0,
        truncate(2.0*Normal(mu=0.0, sigma=1.0) + 1.0, lb=-1.0, ub=1.0),
        order_statistic(truncate(2.0*Normal(mu=0.0, sigma=1.0) + 1.0, lb=-1.0, ub=1.0), r=3.0, n=10.0),
    ],
    weights=array([0.25, 0.25, 0.25, 0.25]),
)
```

#### Additional information
<!--Any additional information you think is important.-->
It would be nice to have a better formatter for these. I played around locally with using `black.format_str`  to format them and it works well. It would be absurd to add `black` as a runtime dependency, let alone using a function that isn't even in the public API.  I guess we'd have to write a custom formatter, or just copy and paste the relevant parts of `black`, which is MIT licensed. If it's worth doing we can leave that for the future.

steppi committed:
MAINT: sort dict of parameters instead of sorting later

- dicts have been sorted since Python 3.7, so let's take
advantage of that. This also ensures that transformed
distributions have their parameters appended after the
original distribution parameters in the __repr__

steppi committed:
MAINT: Update reprs of transformed distributions

mdhaber commented:
Looks nice and simple - great! 
Could you replace the failing test, maybe with a simple test of each of these methods?
Also, it sounds like you're happy with `exp` and `log` working the way they do. But is `inv` OK? It breaks the convention that the `repr` is executable. Also, the way I did `pow` was a hack - it might be good to replace that here, too.

steppi commented:
Thanks @mdhaber!

> Could you replace the failing test, maybe with a simple test of each of these methods?

Sure, no problem

> Also, it sounds like you're happy with `exp` and `log` working the way they do. But is `inv` OK? It breaks the convention that the `repr` is executable. Also, the way I did `pow` was a hack - it might be good to replace that here, too.

Yeah, `exp` and `log` are good. Instead of `inv` maybe `f"1.0 / {repr(self._dist)}"` ? Instead of `pow`,  things like `f"{repr(self._dist)}**{val}"`?. That keeps the executable nature. We'll need to make sure stuff gets parenthesized correctly though.

I'll fix the the test and get rid of the custom `__str__`, and then can try to do the other stuff in a follow up.

mdhaber reviewed (commented):
Oops, didn't submit this before.

steppi committed:
MAINT: raise if no __repr__ override transformed dist subclass

steppi committed:
MAINT: remove __str__ from monotonic transform

- just allow fallback to __repr__

steppi committed:
TST: Update __repr__ tests for continuous dists

steppi committed:
MAINT: Improved reprs for pow, div, abs

steppi committed:
TST: Add tests the reprs evaluate to correct dist

steppi committed:
MAINT: fix repr issues to make them all executable

steppi committed:
Remove now unnecessary repr tests

steppi committed:
MAINT: Fix shifted scaled repr calculation for ndim > 0

steppi committed:
Tst: Fix test_executable for array case

steppi commented:
OK, I fixed that failing test and changed the `repr`s to make everything actually executable. I also added a test to ensure the `repr`s are executable and `eval`ing them gives the correct distribution. There's also a bug I needed to fix for the `repr` of shifted scaled distributions in the array case.

steppi committed:
MAINT: noqa imports that appear unused since used in eval [lint only]

steppi commented:
I'll submit a `lint only` fix with the `noqa`s if everything else settles.

mdhaber commented:
Cool; I just checked that this works well with the recent improvements to `make_distribution`.

```python3
import numpy as np
import matplotlib.pyplot as plt
from scipy import stats
Uniform = stats.make_distribution(stats.uniform)

rng = np.random.default_rng(84954561)

X = Uniform()  # Uniform()
Y = X*5 + 2  # 5.0*Uniform() + 2.0
Z = stats.truncate(Y, 3, 4)  # truncate(5.0*Uniform() + 2.0, lb=3.0, ub=4.0)
W = 1 / Z  # 1/(truncate(5.0*Uniform() + 2.0, lb=3.0, ub=4.0))
Q = stats.exp(W)  # exp(1/(truncate(5.0*Uniform() + 2.0, lb=3.0, ub=4.0)))

x = rng.uniform(size=1000000)
y = x*5 + 2
z = y[(y > 3) & (y < 4)]
w = 1/z
q = np.exp(w)

Q.plot()
plt.hist(q, bins=np.linspace(*Q.support(), 50), density=True)
plt.show()
```
![image](https://github.com/user-attachments/assets/d553c738-4606-496e-81fe-c56c6d9b9637)

<details>
I tried going one step further with `order_statistic`, but executing that function led to an error, which I'll have to fix. Something to do with passing too many arguments to a function; looks easy.
</details>

mdhaber commented on scipy/stats/tests/test_continuous.py:
```diff
@@ -1560,17 +1556,68 @@ def test_generate_domain_support(self):
+            abs(U),
+            stats.exp(U),
+            stats.log(1 + U),
+            np.array([1.0, 2.0])*U + np.array([2.0, 3.0]),
```
Ah, I recently noticed this while working on `marray` but I see it applies here, too.
NumPy arrays define `__mul__`, and the result of this operation is:

```python3
np.array([1.0, 2.0])*U + np.array([2.0, 3.0]),
# array([1.0*_Uniform(a=0.0, b=1.0) + 2.0, 2.0*_Uniform(a=0.0, b=1.0) + 3.0],
#       dtype=object)
```

That's unfortunate, because it should really work like:
```python3
U*np.array([1.0, 2.0]) + np.array([2.0, 3.0]),
# ([1. 2.]*_Uniform(a=0.0, b=1.0) + [2. 3.],)
```
In this test, I'd suggest doing the latter - although something needs adjustment because the arrays look like they have been rendered with `__str__` rather than `__repr__`. 

When that's done, rather than having a separate `_get_fingerprint`, you should just be able to call `sample`.

Fixing the asymmetry of array operators is a different story...

steppi replied:
Done, this worked.

mdhaber commented on scipy/stats/tests/test_continuous.py:
```diff
@@ -1560,17 +1556,68 @@ def test_generate_domain_support(self):
+            np.array([1.0, 2.0])*U + np.array([2.0, 3.0]),
+        ]
+    )
+    def test_executable(self, dist):
```
Nice test!

mdhaber commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -4601,13 +4612,13 @@ def __init__(self, X, /, *args, g, h, dh, logdh=None,
-    def __str__(self):
-        return f"{self._g_name}({str(self._dist)})"
+    def __repr__(self):
+        return self.__repr
```
I didn't immediately understand this `self.__repr`, but I see you're just computing once and returning each time after.

I am not too concerned about execution time of `repr`, so if you're willing, consider just:
```
self._repr_pattern = repr_pattern or f"{g.__name__}(...)"
```
above, and here:
```
return self._repr_pattern.replace("...", repr(self._dist))
```
(And at some point we'll rename `self._dist` to `self._X` or something.)

mdhaber reviewed (commented):
Very close! Thanks for this!

steppi committed:
TST: Get rid of unnecessary fingerprint function

- it turns out this complexity was only needed to have the test pass
despite there being a bug

steppi committed:
MAINT: Ensure repr is used instead of str for nested reprs

steppi committed:
BUG: Set array priority to 1 to get proper behavior

steppi committed:
MAINT: remove stale comment

steppi committed:
TST: Add test for repr with float32 params

steppi committed:
MAINT: Remove dtype and shape from distribution reprs

steppi committed:
MAINT: Avoid hiding information in reprs

- later we'll implement str and can hide the info there

steppi committed:
TST: Update reprs in __repr__ test

steppi committed:
MAINT: Be more aggressive for when to summarize arrays

- avoids overly long reprs

steppi committed:
TST: Add tests that reprs don't get too long

steppi committed:
BUG: fix string pattern conflict with summarized arrays

steppi committed:
TST: Add test case to demonstrate bug

steppi committed:
BUG: Fix bug in loc scale repr

steppi commented on scipy/stats/_distribution_infrastructure.py:
```diff
@@ -3890,10 +3897,10 @@ def _support(self, loc, scale, sign, **params):
 
     def __repr__(self):
         result =  f"{self.scale}*{repr(self._dist)}"
```
No longer neded due to `__array_priority__`

steppi committed:
MAINT: respond to review comment

steppi commented:
OK, so in addition to responding to the comments here, I've also

1. Used the `repr` throughout when building up the `repr` from constituents, so something like
`result =  f"{repr(self.scale)}*{repr(self._dist)}"` instead of `result =  f"{self.scale}*{repr(self._dist)}"`. This ensures the `repr`s remain executable and when there are arrays inside, or NumPy dtypes other than `float64`.

2. Removed `shape` and `dtype` from the `repr` of distributions, since these are not accepted by `__init__`, so they make the `repr` non-executable. By using the `repr`s of internal arrays as mentioned before, we're able to provide the same info to the user without these.

3. Removed code that removed parameter values from the reprs when values were arrays with more than 3 elements. Instead, I've used the `np.printoptions` context manager to be more aggressive in deciding when to summarize arrays. I've added tests to check this works to ensure `repr`s don't get too long.

steppi committed:
Merge branch 'main' into stats-transform-repr

steppi committed:
TST: Update some repr tests

steppi committed:
TST: Add test demonstrating incorrect repr behavior for shift scale

steppi committed:
BUG: Fix case in repr where we hide zero array that causes type promotion

steppi commented:
I fixed one more problem in the `repr` for `ShiftedScaledDistribution` and added a test case. It involved hiding a zero array for `loc` in cases where `loc` needs to be promoted to match `scale`.

steppi committed:
Merge branch 'main' into stats-transform-repr

steppi committed:
TST: Make repr tests work for numpy < 2

steppi commented:
Ah, tests are failing for `numpy < 2` because the `repr`s didn't include the `dtype` back then. I fixed it by skipping the impacted executable test cases and branching on numpy version for the comparisons against a reference.

ev-br commented:
Elsewhere, what we did was we used the numpy 2 repr (`np.float64(2.0)`) and ignored doctests failing on numpy 1.x.

steppi committed:
TST: fix careless error in test

steppi committed:
Merge branch 'main' into stats-transform-repr

steppi commented:
> Elsewhere, what we did was we used the numpy 2 repr (`np.float64(2.0)`) and ignored doctests failing on numpy 1.x.

Makes sense. Thanks for the reminder to check that we don't have any doctests that will break on numpy 1.x here, since we know from #22027 that the doctests aren't running in CI. Looks like we're in the clear because the distribution `repr`s don't appear in any doctests

mdhaber committed:
MAINT: fix __all__, make _repr_pattern private

mdhaber commented:
I made `_repr_pattern` private and fixed `__all__` of `_distribution_infrastructure` (something I noticed recently).

Oops, I didn't see what was happening, but I would have suggested just skipping the `repr` tests pre-NumPy 2.0 rather than branching all those cases.  : /  Just another reason to keep up with SPEC 0....

 I'll merge when tests pass. Thanks @steppi!

steppi commented:
Thanks @mdhaber

> Oops, I didn't see what was happening, but I would have suggested just skipping the `repr` tests pre-NumPy 2.0 rather than branching all those cases. : / Just another reason to keep up with SPEC 0....

OK, yeah, that probably would have been a better idea.
